lib/active_admin/pundit_adapter.rb

```diff
@@ -46,7 +46,9 @@ def retrieve_policy(subject)
     def format_action(action, subject)
       # https://github.com/varvet/pundit/blob/master/lib/generators/pundit/install/templates/application_policy.rb
       case action
+      when Auth::NEW then :new?
```

Comment:
I believe I've figured out why the test is failing. `pundit_adapter_spec.rb:54` calls `authorized?(:create, announcement_post)`, which runs through this case statement with the `action` of `:create`, which matches on both `Auth::NEW` and `Auth::CREATE`. It takes the first match here and runs the `new?` method instead of the `create?` method.

It would work fine if active admin and its tests always called `authorized?` with the `Auth::METHOD` constant format, but it seems interchanged, at least in the tests.

I'd love input on how to resolve this?

Reply:
Hei @timwis!

I think we basically want to fallback to the default case branch, the one falling back to `#{action}?`, so I think removing the conditions where you want just that might do the trick?

```diff
diff --git a/lib/active_admin/pundit_adapter.rb b/lib/active_admin/pundit_adapter.rb
index d32f14504..c4d24570d 100644
--- a/lib/active_admin/pundit_adapter.rb
+++ b/lib/active_admin/pundit_adapter.rb
@@ -46,10 +46,6 @@ module ActiveAdmin
     def format_action(action, subject)
       # https://github.com/varvet/pundit/blob/master/lib/generators/pundit/install/templates/application_policy.rb
       case action
-      when Auth::NEW then :new?
-      when Auth::CREATE then :create?
-      when Auth::EDIT then :edit?
-      when Auth::UPDATE then :update?
       when Auth::READ then subject.is_a?(Class) ? :index? : :show?
       when Auth::DESTROY then subject.is_a?(Class) ? :destroy_all? : :destroy?
       else "#{action}?"
```

Reply:
Thanks @deivid-rodriguez, good idea! I tried this, but because I used backwards-compatible constant-symbol mappings in `authorization_adapter.rb`, it ended up checking the wrong action anyway. `authorization_adapter.rb` set `NEW = :create` and `EDIT = :update` for backwards compatibility. This section was intended to override that for pundit.

But apparently [CanCan considers `:new` and `:create` to be aliases](https://github.com/CanCanCommunity/cancancan/blob/develop/docs/Action-Aliases.md) (and edit/update), so the backwards compatibility is already covered for CanCan — I've added a test to verify that.

The only thing it doesn't cover is backwards compatibility for custom authorization adapters.

I've just pushed changes (including a few minor improvements to the tests for consistency). The unit tests are passing now, but the cucumber tests fail because their custom authorization adapters don't restrict `:new`.

Need to think of a better way to provide backwards compatibility which still allows this to work 🤔

Reply:
The options I've thought of so far are:

1. Leave `NEW = :create` in the AuthorizationAdapter. Then, in the PunditAdapter, somehow check whether the `action` passed to `authorized?` is a constant, and what the constant was. I'm thinking like you might check the name of a class instance's class, but my guess is this isn't possible in Ruby, and it probably just passes the value (`:create`)
2. Have discrete `NEW = :new` in the AuthorizationAdapter. Then, in the AuthorizationAdapter class, somehow add a method for mapping/normalising the actions before passing them to `authorized?`. By default, `:new` gets mapped to `:create` as if it's an alias, but in the pundit adapter, we don't do that (we'd extend the mapping/normalising method to be no-op). Having poked around, though, I'm not sure this is possible either as ActiveAdmin just instantiates the class and calls its `authorized?` method with global constants.
3. Go through the ActiveAdmin codebase and swap out the use of `ActiveAdmin::Auth::<action>` with something that can be overridden by our adapter subclass. Perhaps something like `ActiveAdmin.application.namespace(:admin).authorization_adapter.class::NEW`, which would allow us to set `NEW = :create` in the parent AuthorizationAdapter class but override it with `NEW = :new` in the PunditAdapter subclass.

So far option 3 seems the most promising! Will try it out tomorrow morning. Let me know if you've any other ideas.

Reply:
Just thought of a fourth option:

```diff
  def format_action(action, subject)
    case action
-   when Auth::CREATE then :create?
+   when Auth::CREATE then subject.is_a?(Class) ? :new? : :create?
-   when Auth::UPDATE then :update?
+   when Auth::UPDATE then subject.changed? ? :edit? : :update?
    when Auth::READ then subject.is_a?(Class) ? :index? : :show?
    when Auth::DESTROY then subject.is_a?(Class) ? :destroy_all? : :destroy?
    else "#{action}?"
    end
  end
```

Just wondering whether `.changed?` covers all edge cases 🤔

Reply:
Okay @deivid-rodriguez I've hit a wall after about 6 hours today. I'm quite confident we could actually get this down to a _one-line change_:

```diff
  def format_action(action, subject)
    case action
-   when Auth::CREATE then :create?
+   when Auth::CREATE then subject.is_a?(Class) ? :new? : :create?
    when Auth::UPDATE then :update?
    when Auth::READ then subject.is_a?(Class) ? :index? : :show?
    when Auth::DESTROY then subject.is_a?(Class) ? :destroy_all? : :destroy?
    else "#{action}?"
    end
  end
```

Because I don't think the use case for this PR really applies to edit/update.

This _looks_ like it should work, given how index/show and destroy_all/destroy work below. Unfortunately, it doesn't: when you browse to the `/new` page, `authorized?` is called with an _instantiated_ resource, rather than the Class.

I've tracked this down to:
- [inherited_resources/actions.rb:#new](https://github.com/activeadmin/inherited_resources/blob/master/lib/inherited_resources/actions.rb#L19) calls the `build_resource` method instead of the `collection` method like it uses for `#index`
- This calls [data_access.rb:build_resource](https://github.com/activeadmin/activeadmin/blob/master/lib/active_admin/resource_controller/data_access.rb#L120), which appears to instantiate the resource

I'm not sure why the authorization call on `/new` would use an instantiated object instead of the class. It uses the class to determine whether to show the "New" button, but an instantiated object to access the page. If there's a way we can get it to use the class when accessing the page, this PR can be incredibly simple!

But I've dug about as far as I can :/ Any idea?

Reply:
Hi @timwis, I'm not too sure about what the problem is.

I looked at the failing specs with the change I suggested before, and the failures seem expected, due to how the policy was setup. The unit spec you added fails because the `create?` method of the policy is correctly invoked, but ends up calling `.admin` on the test double that represents the user, which didn't respond to that. Similarly, some cucumber features failed for the same reason: they were calling `create?` on a record without category, and thus crashing.

Changing the unit spec and the policy as follows gets everything passing:

```diff
diff --git a/lib/active_admin/pundit_adapter.rb b/lib/active_admin/pundit_adapter.rb
index d32f14504..c4d24570d 100644
--- a/lib/active_admin/pundit_adapter.rb
+++ b/lib/active_admin/pundit_adapter.rb
@@ -46,10 +46,6 @@ module ActiveAdmin
     def format_action(action, subject)
       # https://github.com/varvet/pundit/blob/master/lib/generators/pundit/install/templates/application_policy.rb
       case action
-      when Auth::NEW then :new?
-      when Auth::CREATE then :create?
-      when Auth::EDIT then :edit?
-      when Auth::UPDATE then :update?
       when Auth::READ then subject.is_a?(Class) ? :index? : :show?
       when Auth::DESTROY then subject.is_a?(Class) ? :destroy_all? : :destroy?
       else "#{action}?"
diff --git a/spec/support/templates/policies/post_policy.rb b/spec/support/templates/policies/post_policy.rb
index 4b26840fd..9490eaa4e 100644
--- a/spec/support/templates/policies/post_policy.rb
+++ b/spec/support/templates/policies/post_policy.rb
@@ -4,7 +4,7 @@ class PostPolicy < ApplicationPolicy
   end
 
   def create?
-    record.category.name != "Announcements" || user.admin
+    record.category.nil? || record.category.name != "Announcements" || user.admin
   end
 
   def update?
diff --git a/spec/unit/pundit_adapter_spec.rb b/spec/unit/pundit_adapter_spec.rb
index c36e88beb..e02739f52 100644
--- a/spec/unit/pundit_adapter_spec.rb
+++ b/spec/unit/pundit_adapter_spec.rb
@@ -29,7 +29,7 @@ RSpec.describe ActiveAdmin::PunditAdapter do
     let(:application) { ActiveAdmin::Application.new }
     let(:namespace) { ActiveAdmin::Namespace.new(application, "Admin") }
     let(:resource) { namespace.register(Post) }
-    let(:auth) { namespace.authorization_adapter.new(resource, double) }
+    let(:auth) { namespace.authorization_adapter.new(resource, double(:admin => false)) }
     let(:default_policy_klass) { DefaultPolicy }
     let(:default_policy_klass_name) { "DefaultPolicy" }
 ```

I also played around with a simple application with a similar policy that always whitelisted `new?` but had some custom behavior on `create?` and things seemed good.

Can you explain the issue my suggested approach has? As I understand it, my approach should be independent from this `Auth` mapping in `authorization.rb`.

Reply:
@timwis 
> I'm not sure why the authorization call on /new would use an instantiated object instead of the class

My guess is that the controller method "new" instantiate a resource and authorize it before rendering

The workaround here is that if you are using a scoped_collection, your instantiated resource should be compliant with the `:create?` policy, because your scoped_collectin would probably leverage your policy scope (I'm still confused on why it's not the case by default)

TLTR : a user belongs to a company, the create new user would ask for the `new?` policy
but a direct hit on /new would instanciate an user instance within the company scope, and your create? policy which would be something like `my_admin_user.company == new_user_record.company` would be valid

Reply:
btw send a PR to scope new resources build (new and create methods) from auth scopes
https://github.com/activeadmin/activeadmin/pull/6884

So when you hit /new and the policy receives the instantiated resource, even if it hits the :create policy it should do the job, as your resource would be scoped by it (and thus probably make your create? pass). Not fixing the issue we have here but it can ease it a bit depending on use cases.

BTW I think, even if my or is not merged, that it's a good illustration of why we should avoid the changed? calls alongside new? to check if the user tries to create? or "render the new page/form". A new resource can have changed attribute even before being presented to any forms, depending on what you have in your controller#new

Reply:
I think all policies should assume they can be called with an instantiated resource so checking whether the subject of the authorization `is_a?` class is not a good solution. For example, from #6884, we can assume that the `new` action will instantiate a new empty object but with the proper foreign key pointing to its parent already filled, and it makes sense to authorize that object.

@timwis How should we move this forward? I think a first step would be to get this green, then we can figure out what backwards compatibility issues we need to address.

Reply:
Sure thing, will work on it this weekend

Reply:
same here will try to close #6884 this weekend

Reply:
Okay @deivid-rodriguez I've implemented the changes you recommended and tests are passing. Note that, since I removed the backwards compatibility in the constants in `ActiveAdmin::Authorization`` (`NEW` is now `:new` not `:create`), removing those `when` lines probably wasn't necessary, but they were already redundant to begin with since the `else` case covers all four.

In order to get the tests to pass, I had to modify the custom authorisation adapters in the feature specs. It was _these_ that I was worried about: if any users have custom adapters that only handle `:create` and `:update` and then have `else true`. But at least we don't have to worry about CanCan users, per my previous comment's link.

Reply:
Thanks @timwis! Alright, I had another look and I do understand the backwards compatibility concerns better now. How about this approach:

```diff
diff --git a/lib/active_admin/authorization_adapter.rb b/lib/active_admin/authorization_adapter.rb
index 171cd0a2a..7fd9d6421 100644
--- a/lib/active_admin/authorization_adapter.rb
+++ b/lib/active_admin/authorization_adapter.rb
@@ -3,7 +3,9 @@ module ActiveAdmin
   # Default Authorization permissions for Active Admin
   module Authorization
     READ = :read
+    NEW = :new
     CREATE = :create
+    EDIT = :edit
     UPDATE = :update
     DESTROY = :destroy
   end
diff --git a/lib/active_admin/base_controller/authorization.rb b/lib/active_admin/base_controller/authorization.rb
index fff86b96a..b38db2df6 100644
--- a/lib/active_admin/base_controller/authorization.rb
+++ b/lib/active_admin/base_controller/authorization.rb
@@ -35,7 +35,7 @@ module ActiveAdmin
       # @return [Boolean]
       #
       def authorized?(action, subject = nil)
-        active_admin_authorization.authorized?(action, subject)
+        active_admin_authorization.authorized?(action_to_permission(action), subject)
       end
 
       # Authorize the action and subject. Available in the controller
@@ -96,7 +96,11 @@ module ActiveAdmin
       # @return [Symbol] The permission name to use.
       def action_to_permission(action)
         if action && action = action.to_sym
-          Authorization::ACTIONS_DICTIONARY[action] || action
+          if active_admin_authorization.respond_to?(:action_to_permission)
+            active_admin_authorization.action_to_permission(action) || action
+          else
+            Authorization::ACTIONS_DICTIONARY[action] || action
+          end
         end
       end
 
diff --git a/lib/active_admin/orm/active_record/comments/views/active_admin_comments.rb b/lib/active_admin/orm/active_record/comments/views/active_admin_comments.rb
index d900b293e..8af0621fc 100644
--- a/lib/active_admin/orm/active_record/comments/views/active_admin_comments.rb
+++ b/lib/active_admin/orm/active_record/comments/views/active_admin_comments.rb
@@ -35,7 +35,7 @@ module ActiveAdmin
 
           text_node paginate @comments
 
-          if authorized?(ActiveAdmin::Auth::CREATE, ActiveAdmin::Comment)
+          if authorized?(ActiveAdmin::Auth::NEW, ActiveAdmin::Comment)
             build_comment_form
           end
         end
diff --git a/lib/active_admin/resource/action_items.rb b/lib/active_admin/resource/action_items.rb
index 4b439b7ee..45bb83236 100644
--- a/lib/active_admin/resource/action_items.rb
+++ b/lib/active_admin/resource/action_items.rb
@@ -64,7 +64,7 @@ module ActiveAdmin
       # Adds the default New link on index
       def add_default_new_action_item
         add_action_item :new, only: :index do
-          if controller.action_methods.include?("new") && authorized?(ActiveAdmin::Auth::CREATE, active_admin_config.resource_class)
+          if controller.action_methods.include?("new") && authorized?(ActiveAdmin::Auth::NEW, active_admin_config.resource_class)
             localizer = ActiveAdmin::Localizers.resource(active_admin_config)
             link_to localizer.t(:new_model), new_resource_path
           end
@@ -74,7 +74,7 @@ module ActiveAdmin
       # Adds the default Edit link on show
       def add_default_edit_action_item
         add_action_item :edit, only: :show do
-          if controller.action_methods.include?("edit") && authorized?(ActiveAdmin::Auth::UPDATE, resource)
+          if controller.action_methods.include?("edit") && authorized?(ActiveAdmin::Auth::EDIT, resource)
             localizer = ActiveAdmin::Localizers.resource(active_admin_config)
             link_to localizer.t(:edit_model), edit_resource_path(resource)
           end
diff --git a/lib/active_admin/view_helpers/auto_link_helper.rb b/lib/active_admin/view_helpers/auto_link_helper.rb
index 3e1d1506f..f78c350f3 100644
--- a/lib/active_admin/view_helpers/auto_link_helper.rb
+++ b/lib/active_admin/view_helpers/auto_link_helper.rb
@@ -28,7 +28,7 @@ module ActiveAdmin
           authorized?(ActiveAdmin::Auth::READ, resource)
           url_for config.route_instance_path resource, url_options
         elsif config.controller.action_methods.include?("edit") &&
-          authorized?(ActiveAdmin::Auth::UPDATE, resource)
+          authorized?(ActiveAdmin::Auth::EDIT, resource)
           url_for config.route_edit_instance_path resource, url_options
         end
       end
diff --git a/lib/active_admin/views/index_as_table.rb b/lib/active_admin/views/index_as_table.rb
index 948569e41..a653e5f95 100644
--- a/lib/active_admin/views/index_as_table.rb
+++ b/lib/active_admin/views/index_as_table.rb
@@ -376,7 +376,7 @@ module ActiveAdmin
           if controller.action_methods.include?("show") && authorized?(ActiveAdmin::Auth::READ, resource)
             item localizer.t(:view), resource_path(resource), class: "view_link #{options[:css_class]}", title: localizer.t(:view)
           end
-          if controller.action_methods.include?("edit") && authorized?(ActiveAdmin::Auth::UPDATE, resource)
+          if controller.action_methods.include?("edit") && authorized?(ActiveAdmin::Auth::EDIT, resource)
             item localizer.t(:edit), edit_resource_path(resource), class: "edit_link #{options[:css_class]}", title: localizer.t(:edit)
           end
           if controller.action_methods.include?("destroy") && authorized?(ActiveAdmin::Auth::DESTROY, resource)
diff --git a/lib/active_admin/views/pages/index.rb b/lib/active_admin/views/pages/index.rb
index f9f12a7fe..6a49316a6 100644
--- a/lib/active_admin/views/pages/index.rb
+++ b/lib/active_admin/views/pages/index.rb
@@ -111,7 +111,7 @@ module ActiveAdmin
 
         def render_blank_slate
           blank_slate_content = I18n.t("active_admin.blank_slate.content", resource_name: active_admin_config.plural_resource_label)
-          if controller.action_methods.include?("new") && authorized?(ActiveAdmin::Auth::CREATE, active_admin_config.resource_class)
+          if controller.action_methods.include?("new") && authorized?(ActiveAdmin::Auth::NEW, active_admin_config.resource_class)
             blank_slate_content = [blank_slate_content, blank_slate_link].compact.join(" ")
           end
           insert_tag(view_factory.blank_slate, blank_slate_content)
```

By doing this, you can define your own adapter class with its own permission to action mapping instead of the default one, by implementing the `action_to_permission` method. For example, if you want to define separate rules for each action in your , you can do:

```ruby
class PunditAdapterWithPerActionRules < ActiveAdmin::PunditAdapter
  def action_to_permission(action)
    action
  end
end

Reply:
Ah! I was trying something very similar with my #2 suggestion above, but in AuthorizationAdapter instead of Authorization. I like the extension approach, too. I'll play around with it and push a commit when I get it working. Thanks for diving in.